jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,375 @@
+import io.netty.channel.ChannelPromise;
+import jdk.jfr.Category;
+import jdk.jfr.Enabled;
+import jdk.jfr.Event;
+import jdk.jfr.Label;
+import jdk.jfr.Name;
+
+import java.net.SocketAddress;
+
+/**
+ * A {@link ChannelHandler} that creates JFR events for all Netty events/operations using the JDK Flight Recorder using
+ * the jdk.jfr APIs directly.
+ * <p>
+ * Added for benchmark comparisons.
+ */
+@Sharable
+@SuppressWarnings({ "StringConcatenationInsideStringBufferAppend", "StringBufferReplaceableByString" })
+public class JfrEventHandler extends ChannelDuplexHandler {
+    // TODO: Check enablement, can we ever be sure not to enable this?
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+        ChannelRegisteredEvent event = new ChannelRegisteredEvent();
+        event.begin();
+        ctx.fireChannelRegistered();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelUnregistered(ChannelHandlerContext ctx) throws Exception {
+        ChannelUnregisteredEvent event = new ChannelUnregisteredEvent();
+        event.begin();
+        ctx.fireChannelUnregistered();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext ctx) throws Exception {
+        ChannelActiveEvent event = new ChannelActiveEvent();
+        event.begin();
+        ctx.fireChannelActive();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        ChannelInactiveEvent event = new ChannelInactiveEvent();
+        event.begin();
+        ctx.fireChannelInactive();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+        ExceptionCaughtEvent event = new ExceptionCaughtEvent();
+        event.begin();
+        ctx.fireExceptionCaught(cause);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+        UserEventTriggeredEvent event = new UserEventTriggeredEvent();
+        event.begin();
+        ctx.fireUserEventTriggered(evt);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.userEvent = String.valueOf(evt);
```

Comment:
This is not really safe imho... The `evt` may have been released at this point already or modified in another way. You will need to convert it to a string before calling `ctx.fireUserEventTriggered(...)`.

Reply:
What about the ctx.channel(), can you expect that to stay the same?

Reply:
yes... the `Channel` that can be accessed via `ChannelHandlerContext` will never change.